simple-preprocess/simple-preprocess.man
## NAME
  simple-preprocess - minimalistic preprocessor for Lua code.

## SYNOPSIS
  `simple-preprocess [OPTION]... [INPUT-FILE] [OUTPUT-FILE]`

## DESCRIPTION
  This tool allows transforming the contents of a Lua source file to generate new code that is specialized for a certain task. For example, it can be used to build a lookup table for a program before the actual program runs, or compile-out certain bits of code that are only needed for testing. This is done by adding lines called 'directives' in the code that will be executed as a Lua program to build the modified version. Directives start with a `##` or `--##` sequence which must not have any leading characters other than whitespace.
  
  The directives are run in a custom environment that indexes back to _ENV, so the regular environment will be accessible as well as any currently defined global variables. This environment also provides the `spwrite()` function that can be called as a directive. The `spwrite()` function behaves very similar to `print()`, but output is sent to the generated file and no whitespace or tabs are added between arguments. The user can also define local variables to add to the custom environment using the `--local-env` option (see `serialization.serialize()` [here](https://ocdoc.cil.li/api:serialization)).
  
  This code is based on the [Simple Lua Preprocessor](http://lua-users.org/wiki/SimpleLuaPreprocessor). The syntax for the inline directives (using `$()`) could be problematic with some code, so `spwrite()` can be used as an alternative.

## OPTIONS
  Input/output files default to `-` to specify standard input/output respectively.
  
  `-h`, `--help`        display help message and exit
  
  `-v`, `--verbose`     print additional debug information to stdout
  
  `--local-env=STRING`  data to append to local environment when processing input (format should use serialization library)

## EXAMPLES
  `simple-preprocess test.lua test.out.lua --local-env={DEBUG=true}`
  
  Input file `test.lua`:
```lua
-- Access a user-defined variable from --local-env.
##if DEBUG then
local function log(fmt, ...) print(string.format(fmt, ...)) end
##else
local function log() end
##end

-- Access a variable in the global environment.
##if math.type then
##spwrite("print(\"math.type is available, we are probably running Lua 5.3+\")")
##end

-- Alternative syntax using '--##'.
--##spwrite("log(\"%s\", \"sample text\")")

-- Build a lookup table.
local lut = {
##for i = 0, 10 do
  ##spwrite("[", i, "] = ", math.sin(math.pi * i / 10), ",")
##end
}

-- The OS environment variables can also be accessed. Also, spwrite() can be used to generate commented code.
##spwrite("-- Current PATH = ", os.getenv().PATH)
```
  
  
  Generated file `test.out.lua`:
```lua
-- Access a user-defined variable from --local-env.
local function log(fmt, ...) print(string.format(fmt, ...)) end

-- Access a variable in the global environment.
print("math.type is available, we are probably running Lua 5.3+")

-- Alternative syntax using '--##'.
log("%s", "sample text")

-- Build a lookup table.
local lut = {
  [0] = 0.0,
  [1] = 0.30901699437495,
  [2] = 0.58778525229247,
  [3] = 0.80901699437495,
  [4] = 0.95105651629515,
  [5] = 1.0,
  [6] = 0.95105651629515,
  [7] = 0.80901699437495,
  [8] = 0.58778525229247,
  [9] = 0.30901699437495,
  [10] = 1.2246467991474e-16,
}

-- The OS environment variables can also be accessed. Also, spwrite() can be used to generate commented code.
-- Current PATH = /bin:/usr/bin:/home/bin:.
```


simple-preprocess/simple-preprocess.lua
--------------------------------------------------------------------------------
-- Minimalistic Lua preprocessor.
-- 
-- @see file://simple-preprocess/README.md
-- @author tdepke2
--------------------------------------------------------------------------------


local fs = require("filesystem")
local serialization = require("serialization")
local shell = require("shell")

local USAGE_STRING = [[Usage: simple-preprocess [OPTION]... [INPUT-FILE] [OUTPUT-FILE]

Options:
  -h, --help                display help message and exit
  -v, --verbose             print additional debug information to stdout
      --local-env=STRING    data to append to local environment when processing
                            input (format should use serialization library)

For more information, run: man simple-preprocess
]]


-- Check if a file was opened successfully, and raise an error if not.
local function checkFileError(filename, file, err)
  if not file then
    if fs.isDirectory(filename) then
      io.stderr:write("simple-preprocess " .. filename .. ": is a directory\n")
    else
      io.stderr:write("simple-preprocess " .. filename .. ": " .. tostring(err) .. "\n")
    end
    os.exit(2)
  end
end


local args, opts = shell.parse(...)

-- Check command line options, open files, start preprocessor, and write results
-- to output.
local function main()
  local inputFilename = (args[1] == nil and "-" or args[1])
  local outputFilename = (args[2] == nil and "-" or args[2])
  if #args > 2 or opts["h"] or opts["help"] then
    io.write(USAGE_STRING)
    os.exit(0)
  end
  local verbose = (opts["v"] or opts["verbose"])
  local appendEnv = {}
  if opts["local-env"] then
    appendEnv = serialization.unserialize(tostring(opts["local-env"]))
    if type(appendEnv) ~= "table" then
      io.stderr:write("simple-preprocess: provided --local-env argument could not be deserialized to a table value\n")
      os.exit(2)
    end
  end
  
  local inputFile, outputFile, err
  
  -- Attempt to open input file (or use stdin).
  if inputFilename == "-" then
    inputFile, err = io.stdin, "missing stdin"
  else
    inputFilename = shell.resolve(inputFilename)
    inputFile, err = io.open(inputFilename)
  end
  checkFileError(inputFilename, inputFile, err)
  
  -- Attempt to open output file (or use stdout).
  if outputFilename == "-" then
    outputFile, err = io.stdout, "missing stdout"
  else
    outputFilename = shell.resolve(outputFilename)
    outputFile, err = io.open(outputFilename, "w")
  end
  checkFileError(outputFilename, outputFile, err)
  
  -- The spwrite function (simple-preprocess write) behaves very similar to
  -- print(), except that output is written to the processed file and no
  -- whitespace or tabs are inserted between arguments (besides a newline at the
  -- end). This function is available within the preprocessor environment.
  local outputLines, outputLineNum = {}, 1
  local function spwrite(...)
    outputLines[outputLineNum] = string.rep("%s", select("#", ...)):format(...) .. "\n"
    outputLineNum = outputLineNum + 1
  end
  
  -- Add a directive (code for preprocessor to run) to the output. Special case
  -- for spwrite() calls since tab alignment in the output can get a bit funky.
  local processedLines, lineNum = {}, 1
  local function addDirective(line, prefixStart, prefixSize)
    -- If the directive is a call to spwrite(), paste any leading whitespace into the first argument to preserve alignment.
    local spwriteIndexEnd = select(2, string.find(line, "^%s*spwrite%s*%(", prefixStart + prefixSize))
    local leadSpacing = string.sub(line, 1, prefixStart - 1)
    
    if spwriteIndexEnd and #leadSpacing > 0 then
      local endParenthesis = string.find(line, "^%s*%)", spwriteIndexEnd + 1)
      processedLines[lineNum] = leadSpacing .. string.sub(line, prefixStart + prefixSize, spwriteIndexEnd) .. "\"" .. leadSpacing .. (endParenthesis and "\"" or "\", ") .. string.sub(line, spwriteIndexEnd + 1) .. "\n"
    else
      processedLines[lineNum] = leadSpacing .. string.sub(line, prefixStart + prefixSize) .. "\n"
    end
  end
  
  -- Iterate each line in the file. Search for '##' or '--##' sequence that only has leading whitespace.
  local line = inputFile:read()
  while line do
    local firstNumberSigns = string.find(line, "##", 1, true)
    if firstNumberSigns and string.match(line, "^%s*##[^#]") then
      addDirective(line, firstNumberSigns, 2)
    elseif firstNumberSigns and string.match(line, "^%s*%-%-##[^#]") then
      addDirective(line, firstNumberSigns - 2, 4)
    else
      processedLines[lineNum] = string.format("spwrite(%q)\n", line)
    end
    line, lineNum = inputFile:read(), lineNum + 1
  end
  inputFile:close()
  
  if verbose then
    io.write("################# PREPARED FILE ##################\n")
    io.write(table.concat(processedLines))
    io.write("(END)\n\n")
  end
  
  -- Prepare a custom environment to use with load(), so user can define local variables in command line.
  local preprocessorEnv = {}
  for k, v in pairs(appendEnv) do
    preprocessorEnv[k] = v
  end
  preprocessorEnv.spwrite = spwrite
  setmetatable(preprocessorEnv, {
    __index = _ENV
  })
  
  if verbose then
    io.write("################## ENVIRONMENT ###################\n")
    for k, v in pairs(preprocessorEnv) do
      io.write(tostring(k), " -> ", tostring(v), "\n")
    end
    io.write("_G -> ", tostring(preprocessorEnv._G), "\n")
    io.write("\n")
  end
  
  -- Execute the processed lines as a chunk. This writes to outputLines.
  local lineNum = 0
  local fn, result = load(function() lineNum = lineNum + 1 return processedLines[lineNum] end, "=(load)", "t", preprocessorEnv)
  if not fn then
    io.stderr:write("simple-preprocess " .. inputFilename .. ": " .. tostring(result) .. "\n")
    os.exit(2)
  end
  local status, result = pcall(fn)
  if not status then
    io.stderr:write("simple-preprocess " .. inputFilename .. ": " .. tostring(result) .. "\n")
    os.exit(2)
  end
  
  if verbose then
    io.write("################## OUTPUT FILE ###################\n")
    io.write(table.concat(outputLines))
    io.write("(END)\n\n")
  end
  
  for _, line in ipairs(outputLines) do
    outputFile:write(line)
  end
  outputFile:close()
end

main()
